_source/_posts/2021-12-06-spring-data-elasticsearch.md

```diff
@@ -0,0 +1,338 @@
+
+Start by creating a `spring-data-elasticsearch` folder for the project, fetch the application JDL from the [GitHub repository](https://github.com/oktadev/okta-spring-data-elasticsearch-example), and generate the application with JHipster:
+
+```shell
+mkdir spring-data-elasticsearch
+cd spring-data-elasticsearch
+https -d raw.githubusercontent.com/oktadev/okta-spring-data-elasticsearch-example/main/blog-reactive-ms.jdl
+jhipster jdl blog-reactive-ms.jdl
+```
+
+{% include setup/cli.md type="jhipster"
+   loginRedirectUri="http://localhost:8080/login/oauth2/code/oidc,http://localhost:8081/login/oauth2/code/oidc,http://localhost:8761/login/oauth2/code/oidc"
+   logoutRedirectUri="http://localhost:8080,http://localhost:8081,http://localhost:8761" %}
+
+
+The JHipster Registry is also a Spring Cloud Config server, and by default, it is configured with the profiles `dev` and `native`, which means the configuration will be provided from the location `docker-compose/central-server-config`. Update `docker-compose/central-server-config/application.yml` with the OIDC settings to be shared with all microservices. Set the values from the `.okta.env` file created with Okta CLI.
+
+```yml
+spring:
+  security:
+    oauth2:
+      client:
+        provider:
+          oidc:
+            issuer-uri: https://{yourOktaDomain}/oauth2/default
+        registration:
+          oidc:
+            client-id: {clientId}
+            client-secret: {clientSecret}
+```
+
+
+For the experimentation in this tutorial, configure the Kibana interface for Elasticsearch, which will allow visualizations of the Elasticsearch data.
+Edit `docker-compose/docker-compose.yml` and add the Kibana service like this:
+
+```yml
+services:
+...  
+  blog-kibana:
+     image: docker.elastic.co/kibana/kibana:7.13.3
+     ports:
+       - 5601:5601
+     environment:
+       ELASTICSEARCH_URL: http://blog-elasticsearch:9200
+       ELASTICSEARCH_HOSTS: '["http://blog-elasticsearch:9200"]'
+...       
+```
+
+Also add the following additional configuration to the `blog-elasticsearch` service, to expose the ports to the local host for sending APIs requests:
+
+```yml
+services:
+  blog-elasticsearch:
+    ...
+    ports:
+      - 9200:9200
+```
+Remove the Keycloak service at the bottom, as Okta will be used as Identity Provider.
+
+Create each applications' container image:
+
+```shell
+cd blog
+./mvnw -DskipTests -ntp -Pprod verify jib:dockerBuild
+```
+
+```shell
+cd gateway
+./mvnw -DskipTests -ntp -Pprod verify jib:dockerBuild
+```
+
+Run the architecture with Docker Compose:
+
+```shell
+cd ../docker-compose
+docker compose up
```

Comment:
I get an error from the gateway here:

```
docker-compose-gateway-1             | 2022-01-26 21:59:43.862 ERROR 1 --- [           main] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Exception during pool initialization.
docker-compose-gateway-1             | 
docker-compose-gateway-1             | org.postgresql.util.PSQLException: The connection attempt failed.
docker-compose-gateway-1             |  at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:313)
docker-compose-gateway-1             |  at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:51)
docker-compose-gateway-1             |  at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:225)
docker-compose-gateway-1             |  at org.postgresql.Driver.makeConnection(Driver.java:466)
docker-compose-gateway-1             |  at org.postgresql.Driver.connect(Driver.java:265)
docker-compose-gateway-1             |  at com.zaxxer.hikari.util.DriverDataSource.getConnection(DriverDataSource.java:138)
docker-compose-gateway-1             |  at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:364)
docker-compose-gateway-1             |  at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:206)
docker-compose-gateway-1             |  at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:476)
docker-compose-gateway-1             |  at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:561)
docker-compose-gateway-1             |  at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115)
docker-compose-gateway-1             |  at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:112)
docker-compose-gateway-1             |  at liquibase.integration.spring.SpringLiquibase.afterPropertiesSet(SpringLiquibase.java:266)
docker-compose-gateway-1             |  at org.springframework.boot.autoconfigure.liquibase.DataSourceClosingSpringLiquibase.afterPropertiesSet(DataSourceClosingSpringLiquibase.java:46)
docker-compose-gateway-1             |  at tech.jhipster.config.liquibase.AsyncSpringLiquibase.initDb(AsyncSpringLiquibase.java:118)
docker-compose-gateway-1             |  at tech.jhipster.config.liquibase.AsyncSpringLiquibase.afterPropertiesSet(AsyncSpringLiquibase.java:103)
docker-compose-gateway-1             |  at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1863)
docker-compose-gateway-1             |  at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1800)
docker-compose-gateway-1             |  at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:620)
docker-compose-gateway-1             |  at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
docker-compose-gateway-1             |  at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
docker-compose-gateway-1             |  at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
docker-compose-gateway-1             |  at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
docker-compose-gateway-1             |  at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
docker-compose-gateway-1             |  at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:953)
docker-compose-gateway-1             |  at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:918)
docker-compose-gateway-1             |  at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583)
docker-compose-gateway-1             |  at org.springframework.boot.web.reactive.context.ReactiveWebServerApplicationContext.refresh(ReactiveWebServerApplicationContext.java:64)
docker-compose-gateway-1             |  at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:732)
docker-compose-gateway-1             |  at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:414)
docker-compose-gateway-1             |  at org.springframework.boot.SpringApplication.run(SpringApplication.java:302)
docker-compose-gateway-1             |  at com.okta.developer.gateway.GatewayApp.main(GatewayApp.java:69)
docker-compose-gateway-1             | Caused by: java.net.UnknownHostException: gateway-postgresql
docker-compose-gateway-1             |  at java.base/java.net.AbstractPlainSocketImpl.connect(Unknown Source)
docker-compose-gateway-1             |  at java.base/java.net.SocksSocketImpl.connect(Unknown Source)
docker-compose-gateway-1             |  at java.base/java.net.Socket.connect(Unknown Source)
docker-compose-gateway-1             |  at org.postgresql.core.PGStream.createSocket(PGStream.java:238)
docker-compose-gateway-1             |  at org.postgresql.core.PGStream.<init>(PGStream.java:98)
docker-compose-gateway-1             |  at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:100)
docker-compose-gateway-1             |  at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:215)
docker-compose-gateway-1             |  ... 31 common frames omitted
```

If I look at my Docker container for PostgreSQL, I see:

```
PostgreSQL Database directory appears to contain a database; Skipping initialization


2022-01-26 21:58:56.774 UTC [1] FATAL:  database files are incompatible with server

2022-01-26 21:58:56.774 UTC [1] DETAIL:  The data directory was initialized by PostgreSQL version 13, which is not compatible with this version 14.1 (Debian 14.1-1.pgdg110+1).
```

Reply:
I was able to fix this by removing all my Docker images and trying again with the following command:

```
docker rmi $(docker images -q) --force
```